warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandCompositeTerms.java

```diff
@@ -106,6 +113,71 @@ private ExpandCompositeTerms(ShardQueryConfiguration config) {
+     */
+    @SuppressWarnings("unchecked")
+    public static <T extends JexlNode> T expandTermsWithRegex(ShardQueryConfiguration config, T script, ScannerFactory scannerFactory,
+                    MetadataHelper metadataHelper, Set<String> expansionFields) throws InstantiationException, IllegalAccessException, TableNotFoundException {
+        
+        ExpandCompositeTerms visitor = new ExpandCompositeTerms(config);
+        
+        // need to flatten the tree so i get all and nodes at the same level
+        script = TreeFlatteningRebuildingVisitor.flatten(script);
+        
+        if (null == visitor.config.getCompositeToFieldMap()) {
+            QueryException qe = new QueryException(DatawaveErrorCode.DATATYPESFORINDEXFIELDS_MULTIMAP_MISSING);
+            throw new DatawaveFatalQueryException(qe);
+        }
+        
+        T newTree = (T) script.jjtAccept(visitor, new ExpandData());
+        
+        if (newTree instanceof ASTJexlScript) {
+            Set<String> indexedFields = null;
+            Set<String> indexOnlyFields = null;
+            Set<String> nonEventFields = null;
+            if (config.getMinSelectivity() > 0) {
+                try {
+                    indexedFields = metadataHelper.getIndexedFields(config.getDatatypeFilter());
+                    indexOnlyFields = metadataHelper.getIndexOnlyFields(config.getDatatypeFilter());
+                    nonEventFields = metadataHelper.getNonEventFields(config.getDatatypeFilter());
+                } catch (TableNotFoundException te) {
+                    QueryException qe = new QueryException(DatawaveErrorCode.METADATA_ACCESS_ERROR, te);
+                    throw new DatawaveFatalQueryException(qe);
+                }
+            }
+            
+            final ParallelIndexExpansion regexExpansion = new ParallelIndexExpansion(config, scannerFactory, metadataHelper, expansionFields,
+                            config.isExpandFields(), config.isExpandValues(), config.isExpandUnfieldedNegations());
+            try {
+                T expandedTree = (T) regexExpansion.visit((ASTJexlScript) newTree, null);
+                /**
+                 * Don't promote a tree that isn't executable. If that is the case resort to returning the tree created by the composite expansion.
+                 */
+                if (ExecutableDeterminationVisitor.isExecutable(expandedTree, config, indexedFields, indexOnlyFields, nonEventFields, null, metadataHelper)) {
+                    return (T) expandedTree.jjtAccept(visitor, new ExpandData());
+                }
+            } catch (CannotExpandUnfieldedTermFatalException cee) {
+                /**
+                 * Exceptions in the executor of ParallelIndexExpansion wrap other exceptions. In the case where we have a do not perform query optimization
+                 * exception we want to pass that along. Therefore we will check the cause and return the original DoNotPerformOptimizedQueryException.
```

Comment:
The ParallelIndexExpansion wraps execution exceptions with CannotExpandUnfieldedTermFatalExceptions. In the case of an exception while expanding regexes the original cause will be DoNotPerformOptimizedQueryException. If we failed to expand a regex, persist the original DoNotPerformOptimizedQueryException.

Or put another way "Persist the original exception if a regex term failed to expand"

Reply:
Thanks for the comments. For the life of me I cannot understand why I chose such a verbose way to explain a very simple piece of code. Thanks for pointing this out. At least I can laugh at myself.